### ISSUE TITLE
Certificate verification failed against openssl from homebrew (1.0.2a-1)

### ISSUE THREAD

**Issue by joelwurtz** [boot2docker/boot2docker#808]
Hello,

Depending on the version of openssl certificate verification of boot2docker failed, i don't know if this due to a bug in openssl or the way of boot2docker certificate has been generated, here is some log : 

```
➜  ~  openssl version
OpenSSL 0.9.8zc 15 Oct 2014
➜  ~  openssl s_client -CAfile /Users/joelwurtz/.boot2docker/certs/boot2docker-vm/ca.pem -cert /Users/joelwurtz/.boot2docker/certs/boot2docker-vm/cert.pem -key /Users/joelwurtz/.boot2docker/certs/boot2docker-vm/key.pem -connect boot2docker:2376 -tls1 
CONNECTED(00000003)
depth=1 /O=Boot2Docker
verify return:1
depth=0 /O=Boot2Docker
verify return:1
```

```
➜  ~  /usr/local/opt/openssl/bin/openssl version
OpenSSL 1.0.2a 19 Mar 2015
➜  ~  /usr/local/opt/openssl/bin/openssl s_client -CAfile /Users/joelwurtz/.boot2docker/certs/boot2docker-vm/ca.pem -cert /Users/joelwurtz/.boot2docker/certs/boot2docker-vm/cert.pem -key /Users/joelwurtz/.boot2docker/certs/boot2docker-vm/key.pem -connect boot2docker:2376 -tls1

CONNECTED(00000003)
depth=0 O = Boot2Docker
verify error:num=18:self signed certificate
verify return:1
depth=0 O = Boot2Docker
verify error:num=21:unable to verify the first certificate
verify return:1
...

**Comment by pwnall** [comment]
+1

I ran into the same issue. This is causing me trouble when trying to use the docker API from a Ruby app. Here are the openssl commands that I used. They're similar to the ones above, but print more debugging information and use DOCKER_ env vars.

```bash
openssl s_client -showcerts -connect ${DOCKER_HOST/tcp:\/\//} \
    -key $DOCKER_CERT_PATH/key.pem \
    -cert $DOCKER_CERT_PATH/cert.pem \
    -CAfile $DOCKER_CERT_PATH/ca.pem -tls1
```

```bash
/usr/local/Cellar/openssl/1.0.2a-1/bin/openssl \
    s_client -showcerts -connect ${DOCKER_HOST/tcp:\/\//} \
    -key $DOCKER_CERT_PATH/key.pem \
    -cert $DOCKER_CERT_PATH/cert.pem \
    -CAfile $DOCKER_CERT_PATH/ca.pem -tls1
```

**Comment by aanand** [comment]
I can reproduce this too.

Working OpenSSL versions:

```
OpenSSL 0.9.8zf 19 Mar 2015
OpenSSL 1.0.1f 6 Jan 2014
OpenSSL 1.0.1j 15 Oct 2014
```

Non-working:

```
OpenSSL 1.0.2a 19 Mar 2015
OpenSSL 1.0.2d 9 Jul 2015
```

`0.9.8zf` is the OS X system `openssl` - all others are installed with Homebrew.

**Comment by aanand** [comment]
Note that I **cannot** reproduce this error against a Boot2Docker VM provisioned with `docker-machine` - it only happens against a VM provisioned with the `boot2docker` command.

This point came up a few weeks ago with Compose and I discussed it with @ehazlett [here](https://github.com/docker/compose/issues/890#issuecomment-106306690), but we couldn't at the time think of a reason why one would work but not the other.

**Comment by aanand** [comment]
I've pasted the output of the command failing with a boot2docker VM and succeeding with a docker-machine VM here:

https://gist.github.com/aanand/3d865623481ba8ae66ee

**Comment by tdsmith** [comment]
I wonder if the problem is that the Issuer and Subject DNs are the same for the boot2docker certs.

Note
```
Server certificate
subject=/O=Boot2Docker
issuer=/O=Boot2Docker
```
vs
```
Server certificate
subject=/O=dev
issuer=/O=aanand
```

**Comment by tdsmith** [comment]
FWIW the boot2docker certs are generated here: https://github.com/boot2docker/boot2docker/blob/30800b21799f3aedba2150c729aeafc13baad8d9/rootfs/rootfs/usr/local/etc/init.d/docker#L44

using this program: https://github.com/SvenDowideit/generate_cert/blob/master/generate_cert.go

**Comment by ehazlett** [comment]
Would you mind running the following:

CA: `openssl x509 -in /Users/joelwurtz/.boot2docker/certs/boot2docker-vm/ca.pem -noout -text`
Server: `openssl x509 -in server.pem -noout -text`
Client: `openssl x509 -in /Users/joelwurtz/.boot2docker/certs/boot2docker-vm/cert.pem -noout -text`

I'm not sure where the "server.pem" file may be.  Perhaps in the boot2docker-vm dir or on the VM in /var/lib/boot2docker/tls/server.pem

Thanks

**Comment by aanand** [comment]
> I'm not sure where the "server.pem" file may be. Perhaps in the boot2docker-vm dir or on the VM in /var/lib/boot2docker/tls/server.pem

That's where it was for me, yes. Output of all 3 here: https://gist.github.com/aanand/d57d22ff198a66d7c354

**Comment by posita** [comment]
Needs SvenDowideit/generate_cert#10. See https://github.com/docker/compose/issues/890#issuecomment-126854371 et seq. for relevant discussion (which probably should have happened here).

**Comment by posita** [comment]
PR with fix submitted as boot2docker/boot2docker#1029.

**Comment by nagarajanewlife** [comment]
verify error:num=18:self signed certificate.
